### ISSUE TITLE
API for registering fields to a schema

### ISSUE THREAD

**Issue by danielbachhuber** [WP-API/WP-API#875]
Something like:

`register_api_field( $resource_name, $attribute_name, $get_callback, $update_callback, $schema );`

Or whatever arguments we're keen on.

**Comment by rmccue** [comment]
Should probably be an array for the callbacks, for forwards-compat if we need to add extra flags.

```php
register_api_field( $resource_name, $attribute_name, array(
    'get_callback' => $get_callback,
    'update_callback' => $update_callback,
    'schema' => $schema,
) );
```

Do we potentially need a `capability_callback` here?

**Comment by danielbachhuber** [comment]
We could also use this API to register many internal fields too.

**Comment by TimothyBJacobs** [comment]
+1 million. #560 might be worth looking at again now. I'd planned on contributing, but then life happened. But it seems like explicit versioning would work hand in hand with a schema. Especially for things like deprecating fields and the like.

**Comment by joehoyle** [comment]
### Implementation Questions

So, it seems this would be best off all implemented in the `WP_JSON_Controller` base class for the most part, as it shouldn't need to interact with the rest of the server level stuff. The only part we need to solve there is how to get the extra fields added to the params array of the endpoint. 

Things like the permissions checks, update callbacks etc can all be done automatically from the `WP_JSON_Controller` classes. Currently our implementation for subclasses is override `get_item` / `get_item_schema` etc, which are then directly called by the route callback. The issue here is that we don't get to implement any DRY middleware for getting schema, checking callbacks etc. I'd rather not have to handle applying the extra fields to the schema, checking permissions callbacks of extra fields etc in all the JSON Controllers, rather just have it implemented once in the base controller and have that be called.

The best way of doing that I can see if either enforce `parent::get_item_schema()` in the subclasses, or introduce a middleware function in the base controller `item_schema_callback` (and for all the other methods invoked via an endpoint callback.) This function would apply any generic middleware it wanted, such as adding in the extra registered fields to the schema, and then internally call `get_item_schema`.

It seems also that the extra fields could be stored on the specific `WP_JSON_Controller` such as `WP_JSON_Posts_Controller::extra_fields` so though we could have a wrapper function, it would invoke `static WP_JSON_Posts_Controller::add_extra_field( $field, ... )` this would need to be static  as we don't have singleton / shared instances. Either that, or registered an additional field would need to be stored in some global until the time it's needed.

We also have to deal with the fact that one subclass of `WP_JSON_Controller` can handle multiple object types (`post`/`page`) for example, so whatever we do in the implementation will probably need subclasses to override that.

Can discuss in more detail in the 16th Feb Meeting.

**Comment by danielbachhuber** [comment]
From our call on Monday, this conceivably could replace the meta endpoints entirely, solving all of their associated problems.

**Comment by danielbachhuber** [comment]
@joehoyle any ideas on a clean, sane way we could:

* Have the schema endpoints automatically registered?
* Pull out resource arguments for registration?
* Add named route to the index? #577 

.. when the server detects we're using our special controller class?

**Comment by joehoyle** [comment]
Initial PR / experimentation in https://github.com/WP-API/WP-API/pull/927

**Comment by rachelbaker** [comment]
#927 Just needs a refresh, and it would be nice to have the ability to register api fields in Beta 1.  This is a NICE to have, not a blocker.

**Comment by danielbachhuber** [comment]
This is beta2 material

**Comment by joehoyle** [comment]
This was merged in #927 - I'm going to close this out, and we can open any specific items a new.
